### PR TITLE
Run tests for integrations  in parallel

### PR DESCRIPTION
**Pull request by mtojek**
## What does this PR do?

This PR modifies the Jenkins pipeline configuration to dynamically create stages and run integration tests in parallel. I'll leverage from this one while working on the https://github.com/elastic/integrations/issues/370 , so I won't break the pipeline for other packages (at some point in time Metricbeat will start publishing metrics under a new key).

Still TODOs:

- [x] Don't spawn the Elastic stack if there is no tests defined
- [x] Build integrations for UpdatePackageStorage

## How to test this PR locally

The CI builder will verify it.

## Related issues

- Resolves https://github.com/elastic/integrations/issues/297

### CODE REVIEW

**Comment by mtojek** [comment]
@kuisathaverat I'm trying to build and test integrations in parallel, but unfortunately the pipeline fails with the following error:

```
Unhandled arch in NODE_LABELS: master
— Error signal
<1s
```

For reference:
https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/PR-421/8/pipeline/167

I'm not quite sure if it started failing after the weekend outage. I would appreciate if you can provide me some guidance here.

**Comment by mtojek** [comment]
I think I went further, just need to select the correct host labels.

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -47,44 +46,44 @@ pipeline {
...
             // For reference: https://issues.jenkins.io/browse/JENKINS-52750
             findFiles()?.findAll{ !it.name.endsWith('@tmp') }?.collect{ it.name }?.sort()?.each {
-              dir("${it}"){
-                sh(label: "Test integration: ${it}", script: '''
-                eval "$(../../build/elastic-package stack shellinit)"
-                ../../build/elastic-package test -v --report-format xUnit --report-output file
-                ''')
+              integrations["${it}"] = {
+                env.NODE_LABELS = "master linux x86_64"
+                node {
```
Comment (kuisathaverat) [inline .ci/Jenkinsfile:55]:
```suggestion
                node('ubuntu-18 && immutable') {
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -47,44 +46,44 @@ pipeline {
...
             // Include hack to skip temporary files with "@tmp" suffix.
             // For reference: https://issues.jenkins.io/browse/JENKINS-52750
             findFiles()?.findAll{ !it.name.endsWith('@tmp') }?.collect{ it.name }?.sort()?.each {
-              dir("${it}"){
-                sh(label: "Test integration: ${it}", script: '''
-                eval "$(../../build/elastic-package stack shellinit)"
-                ../../build/elastic-package test -v --report-format xUnit --report-output file
-                ''')
+              integrations["${it}"] = {
+                env.NODE_LABELS = "master linux x86_64"
```
Comment (kuisathaverat) [inline .ci/Jenkinsfile:54]: not sure what this for

```suggestion
                env.NODE_LABELS = "master linux x86_64"
```

File: .ci/Jenkinsfile (hunk 1 above)
Comment (mtojek) [reply .ci/Jenkinsfile:55]: Let me try this one...

File: .ci/Jenkinsfile (hunk 2 above)
Comment (mtojek) [reply .ci/Jenkinsfile:54]: Firstly it complained about arch, then os and then missing Go runtime. I will try your suggestion with `node('ubuntu-18 && immutable')`

**Comment by mtojek** [comment]
Ok, it looks like I hit this bug: https://issues.jenkins.io/browse/JENKINS-55438

**Comment by mtojek** [comment]
Sample pipeline with a failed test (on purpose): https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/PR-421/26/pipeline/184/

**Comment by mtojek** [comment]
jenkins run the tests please

**Comment by mtojek** [comment]
UpdatePackageStorage tested (works): https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/PR-421/31/pipeline/3148

**Comment by mtojek** [comment]
@ycombinator This PR is finally ready for review. Please take a look at it if you're free.

**Review by ycombinator — APPROVED** [review]
This is amazing - comparing the [latest CI run for this PR](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/PR-421/32/pipeline/195) with a [`master` branch CI run](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/master/19/pipeline/123), looks like we save almost 3 minutes overall! And that's despite the fact that we now have to start the Elastic Stack on each parallel node (if necessary), which takes about 2m.

It would be nice if someday in the future we could share the Elastic Stack amongst the parallel steps but I don't know enough about Jenkins networking to know if this would be possible or how complex it might be. Anyway, this is something for later. Already with this PR, as we add more integrations packages, we're going to start seeing benefits!

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -47,44 +45,40 @@ pipeline {
...
+                node('ubuntu && immutable') {
+                  stage("${it}: check") {
+                    deleteDir()
+                    gitCheckout(basedir: "${BASE_DIR}")
+                    useElasticPackage()
+                    try {
+                      dir("${BASE_DIR}/packages/${it}") {
+                        sh(label: "Check integration: ${it}", script: '../../build/elastic-package check -v')
+                        def r = sh(label: "Find test folders", script: 'find . -d  -name test | grep .', returnStatus: true)
+                        if (r == 0) {
```
Comment (ycombinator) [inline .ci/Jenkinsfile:62]: Nit: IDK enough Groovy but if you could invert this condition and [return early](https://medium.com/swlh/return-early-pattern-3d18a41bba8) that would be nice for extra readability but also not absolutely required at this time.

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -47,44 +45,40 @@ pipeline {
...
+              integrations["${it}"] = {
+                node('ubuntu && immutable') {
+                  stage("${it}: check") {
+                    deleteDir()
+                    gitCheckout(basedir: "${BASE_DIR}")
+                    useElasticPackage()
+                    try {
+                      dir("${BASE_DIR}/packages/${it}") {
+                        sh(label: "Check integration: ${it}", script: '../../build/elastic-package check -v')
+                        def r = sh(label: "Find test folders", script: 'find . -d  -name test | grep .', returnStatus: true)
```
Comment (ycombinator) [inline .ci/Jenkinsfile:61]: Maybe add a comment here explaining why you're doing this? I know you explained this in the PR description but a comment would help anyone in the future looking at this code without having to look up the PR for it.

File: .ci/Jenkinsfile (hunk 3 above)
Comment (mtojek) [reply .ci/Jenkinsfile:62]: It's a short code block, let's leave it as is, but thanks for the suggestion!

File: .ci/Jenkinsfile (hunk 4 above)
Comment (mtojek) [reply .ci/Jenkinsfile:61]: Fixed

**Review by ycombinator — APPROVED** [review]
LGTM.

**Pull request merged**